File: scripts/extract_sessions.py
#!/usr/bin/env python3
"""
extract_sessions.py
从 Claude Code 的 .jsonl 会话文件中提取对话骨架

Usage:
    python extract_sessions.py --list                  # 列出所有项目
    python extract_sessions.py --project <hash_or_path> [--since 7d]
"""

import json
import os
import sys
import argparse
import re
from pathlib import Path
from datetime import datetime, timedelta
from typing import Optional


# ── 跨平台路径 ─────────────────────────────────────────────────────────────────
def get_claude_projects_dir() -> Path:
    """返回 ~/.claude/projects 的跨平台路径"""
    if sys.platform == "win32":
        base = Path(os.environ.get("APPDATA", Path.home()))
        return base / "Claude" / "projects"
    else:
        return Path.home() / ".claude" / "projects"


# ── 项目发现 ──────────────────────────────────────────────────────────────────
def discover_projects(projects_dir: Path) -> list[dict]:
    """
    扫描 projects 目录，尝试从 .jsonl 里的 cwd 字段反推项目路径
    返回 [{ hash, path_hint, latest_mtime, session_count }]
    """
    if not projects_dir.exists():
        return []

    results = []
    for proj_dir in sorted(projects_dir.iterdir()):
        if not proj_dir.is_dir():
            continue

        jsonl_files = sorted(proj_dir.glob("*.jsonl"), key=lambda f: f.stat().st_mtime)
        if not jsonl_files:
            continue

        latest_mtime = max(f.stat().st_mtime for f in jsonl_files)
        path_hint = _guess_project_path(jsonl_files)

        results.append({
            "hash": proj_dir.name,
            "path_hint": path_hint or "(unknown)",
            "latest_mtime": latest_mtime,
            "latest_dt": datetime.fromtimestamp(latest_mtime),
            "session_count": len(jsonl_files),
        })

    # 按最近活跃排序
    results.sort(key=lambda x: x["latest_mtime"], reverse=True)
    return results


def _guess_project_path(jsonl_files: list[Path]) -> Optional[str]:
    """从 .jsonl 文件的第一条记录里尝试读取 cwd"""
    for f in reversed(jsonl_files):  # 最新的文件更可能有 cwd
        try:
            with open(f, "r", encoding="utf-8") as fp:
                for line in fp:
                    line = line.strip()
                    if not line:
                        continue
                    record = json.loads(line)
                    cwd = record.get("cwd") or record.get("workingDirectory")
                    if cwd:
                        return cwd
        except Exception:
            continue
    return None


def print_project_list(projects: list[dict]):
    """打印可交互的项目列表"""
    print("\n📂 Claude Code 项目列表（按最近活跃排序）\n")
    print(f"  {'#':<4} {'最近活跃':<22} {'会话数':<6} 项目路径")
    print("  " + "─" * 75)
    for i, p in enumerate(projects):
        dt_str = p["latest_dt"].strftime("%Y-%m-%d %H:%M")
        print(f"  {i+1:<4} {dt_str:<22} {p['session_count']:<6} {p['path_hint']}")
    print()


# ── 消息提取 ──────────────────────────────────────────────────────────────────
MIN_TEXT_LENGTH = 20  # 短于这个字符数的消息丢弃

# 注入型内容前缀：skill / hook / system-reminder 注入的伪 user 消息，应当过滤
NOISE_PREFIXES = (
    "Base directory for this skill",
    "<system-reminder",
    "<SUBAGENT-STOP",
    "<command-message",
    "<command-name",
    "<local-command",
    "<task-notification",
    "[Request interrupted",
    "Caveat:",
    "This session is being continued",
)


def is_noise(text: str) -> bool:
    """判断一条 user 文本是否为注入型噪声（skill/hook/系统提示），而非真实用户发言"""
    stripped = text.lstrip()
    return any(stripped.startswith(p) for p in NOISE_PREFIXES)

def extract_text_content(content) -> Optional[str]:
    """
    从 message content 字段中提取纯文本。
    content 可能是 str，或 list of blocks。
    跳过含 tool_use / tool_result / thinking 类型的块。
    """
    if isinstance(content, str):
        return content.strip() if len(content.strip()) >= MIN_TEXT_LENGTH else None

    if isinstance(content, list):
        text_parts = []
        has_tool = False
        for block in content:
            if not isinstance(block, dict):
                continue
            btype = block.get("type", "")
            if btype in ("tool_use", "tool_result", "thinking"):
                has_tool = True
                continue
            if btype == "text":
                t = block.get("text", "").strip()
                if t:
                    text_parts.append(t)

        # 如果整条消息只有工具调用，没有文字，丢弃
        combined = "\n".join(text_parts).strip()
        if not combined or len(combined) < MIN_TEXT_LENGTH:
            return None
        return combined

    return None


def parse_timestamp(record: dict) -> Optional[datetime]:
    """尝试解析记录中的时间戳"""
    for key in ("timestamp", "createdAt", "created_at", "ts"):
        val = record.get(key)
        if not val:
            continue
        try:
            if isinstance(val, (int, float)):
                return datetime.fromtimestamp(val / 1000 if val > 1e10 else val)
            dt = datetime.fromisoformat(str(val).replace("Z", "+00:00"))
            # CC 时间戳是 UTC（带 tzinfo）。会话标题用的是本地 mtime，若这里直接
            # strftime 会打印 UTC 墙钟，导致同一条时间线 UTC/本地混用、自相矛盾。
            # 统一转成本地时区再去掉 tzinfo，与会话标题保持一致。
            if dt.tzinfo is not None:
                dt = dt.astimezone().replace(tzinfo=None)
            return dt
        except Exception:
            continue
    return None


def extract_from_jsonl(filepath: Path) -> list[dict]:
    """
    从单个 .jsonl 文件提取对话骨架消息列表
    返回 [{ role, text, timestamp }]
    """
    messages = []
    file_mtime = datetime.fromtimestamp(filepath.stat().st_mtime)

    try:
        with open(filepath, "r", encoding="utf-8") as fp:
            for line in fp:
                line = line.strip()
                if not line:
                    continue
                try:
                    record = json.loads(line)
                except json.JSONDecodeError:
                    continue

                # 当前 CC 格式：每行 type 为 "user"/"assistant"，role/content 嵌在 message 里；
                # 兼容旧格式 type == "message"（顶层 role/content）
                record_type = record.get("type")
                if record_type not in ("user", "assistant", "message"):
                    continue

                # 结构化过滤（比文本前缀匹配更稳健，跨 CC 版本不易失效）：
                # isSidechain=子代理内部对话，会污染主线骨架；isMeta=系统注入的元消息
                if record.get("isSidechain") or record.get("isMeta"):
                    continue

                msg = record.get("message")
                if not isinstance(msg, dict):
                    msg = record

                role = msg.get("role") or record.get("role")
                if role not in ("user", "assistant", "human"):
                    continue

                content = msg.get("content")
                if content is None:
                    continue

                text = extract_text_content(content)
                if not text:
                    continue

                # 过滤 skill/hook 注入到 user 角色里的伪消息
                if role in ("user", "human") and is_noise(text):
                    continue

                ts = parse_timestamp(record) or file_mtime

                messages.append({
                    "role": "human" if role in ("user", "human") else "assistant",
                    "text": text,
                    "timestamp": ts,
                })
    except Exception as e:
        print(f"  ⚠️  读取 {filepath.name} 时出错: {e}", file=sys.stderr)

    return messages


def load_project_sessions(
    proj_dir: Path,
    since: Optional[datetime] = None
) -> list[dict]:
    """
    加载项目下所有会话，返回按时间排序的消息列表
    每条消息额外带 session_id 字段
    """
    jsonl_files = sorted(proj_dir.glob("*.jsonl"), key=lambda f: f.stat().st_mtime)

    if since:
        jsonl_files = [f for f in jsonl_files if datetime.fromtimestamp(f.stat().st_mtime) >= since]

    all_messages = []
    for f in jsonl_files:
        session_msgs = extract_from_jsonl(f)
        if not session_msgs:
            continue
        session_id = f.stem[:8]  # 取 UUID 前8位作为简短标识
        # 会话起点用最早一条消息的真实时间戳，而非文件 mtime——
        # mtime 是"最后修改时刻"，被 resume 过的会话会偏成"恢复时刻"，失真。
        session_start = min(m["timestamp"] for m in session_msgs)
        for msg in session_msgs:
            msg["session_id"] = session_id
            msg["session_start"] = session_start
        all_messages.extend(session_msgs)

    # 全局按时间戳排序
    all_messages.sort(key=lambda m: m["timestamp"])
    return all_messages


# ── 骨架格式化输出 ────────────────────────────────────────────────────────────
def format_skeleton(messages: list[dict], project_path: str, max_chars: int = 0) -> str:
    """
    输出时间线骨架 Markdown。
    max_chars: 单条消息最大字符数，0 表示不截断（默认）。
    注意：骨架是 Step 4 撰文的源材料，最终的分析报告/交付物往往就在最长的那几条里，
    默认截断会静默地给成品质量设天花板，所以默认保留全文。需要精简预览时再传 max_chars。
    """
    lines = [
        f"# 开发过程骨架 — {project_path}",
        f"_生成时间：{datetime.now().strftime('%Y-%m-%d %H:%M')}_",
        f"_共 {len(messages)} 条对话消息_\n",
    ]

    current_session = None
    for msg in messages:
        sid = msg["session_id"]
        if sid != current_session:
            current_session = sid
            lines.append(f"\n## 会话 {sid}  ·  {msg['session_start'].strftime('%Y-%m-%d %H:%M')}\n")

        role_label = "👤 用户" if msg["role"] == "human" else "🤖 Claude"
        ts = msg["timestamp"].strftime("%H:%M")
        text = msg["text"]
        if max_chars and len(text) > max_chars:
            text = text[:max_chars] + "…（已截断）"
        lines.append(f"**{role_label}** `{ts}`\n{text}\n")

    return "\n".join(lines)


# ── 主程序 ────────────────────────────────────────────────────────────────────
def parse_since(s: str) -> Optional[datetime]:
    """解析 --since 参数，如 7d / 30d / 2024-01-01"""
    m = re.match(r"^(\d+)d$", s)
    if m:
        return datetime.now() - timedelta(days=int(m.group(1)))
    try:
        return datetime.fromisoformat(s)
    except Exception:
        return None


def main():
    parser = argparse.ArgumentParser(description="Claude Code 会话提取工具")
    parser.add_argument("--list", action="store_true", help="列出所有项目")
    parser.add_argument("--project", type=str, help="项目 hash 或序号（从 --list 获取）")
    parser.add_argument("--since", type=str, default=None, help="只包含最近N天，如 7d 或 2024-01-01")
    parser.add_argument("--output", type=str, default=None, help="输出文件路径（默认打印到 stdout）")
    parser.add_argument("--max-chars", type=int, default=0, help="单条消息最大字符数，0=不截断（默认）。仅在需要精简预览时设置")
    args = parser.parse_args()

    projects_dir = get_claude_projects_dir()
    projects = discover_projects(projects_dir)

    if not projects:
        print(f"❌ 未找到任何 Claude Code 项目，请确认 {projects_dir} 存在")
        sys.exit(1)

    if args.list or not args.project:
        print_project_list(projects)
        if not args.project:
            print("👉 使用方式：python extract_sessions.py --project <序号或hash> [--since 7d]")
        return

    # 解析 --project 参数
    target = args.project.strip()
    proj_dir = None

    if target.isdigit():
        idx = int(target) - 1
        if 0 <= idx < len(projects):
            proj_dir = projects_dir / projects[idx]["hash"]
        else:
            print(f"❌ 序号 {target} 超出范围（共 {len(projects)} 个项目）")
            sys.exit(1)
    else:
        # 尝试按 hash 或路径模糊匹配
        for p in projects:
            if target in p["hash"] or target in p["path_hint"]:
                proj_dir = projects_dir / p["hash"]
                break
        if not proj_dir:
            print(f"❌ 未找到匹配的项目：{target}")
            sys.exit(1)

    since = parse_since(args.since) if args.since else None
    print(f"\n⏳ 正在读取项目会话...")

    messages = load_project_sessions(proj_dir, since=since)

    if not messages:
        print("⚠️  未提取到任何有效消息（可能被 --since 过滤或文件格式不符）")
        sys.exit(0)

    # 找对应的 path_hint
    matched = next((p for p in projects if (projects_dir / p["hash"]) == proj_dir), {})
    project_path = matched.get("path_hint", proj_dir.name)

    skeleton = format_skeleton(messages, project_path, max_chars=args.max_chars)

    if args.output:
        Path(args.output).write_text(skeleton, encoding="utf-8")
        print(f"✅ 骨架已写入：{args.output}")
    else:
        print(skeleton)


if __name__ == "__main__":
    main()

File: scripts/extract_sessions.js
#!/usr/bin/env node
/*
 * extract_sessions.js
 * 从 Claude Code 的 .jsonl 会话文件中提取对话骨架（Node 版，零依赖）。
 * 与 extract_sessions.py 行为一致——有 Node 没 Python 时用它，反之用 .py。
 *
 * Usage:
 *   node extract_sessions.js --list
 *   node extract_sessions.js --project <序号或hash> [--since 7d] [--output out.md] [--max-chars 0]
 */

const fs = require("fs");
const os = require("os");
const path = require("path");

// ── 跨平台路径 ───────────────────────────────────────────────────────────────
function getClaudeProjectsDir() {
  if (process.platform === "win32") {
    const base = process.env.APPDATA || os.homedir();
    return path.join(base, "Claude", "projects");
  }
  return path.join(os.homedir(), ".claude", "projects");
}

// ── 工具：本地时间格式化（CC 时间戳为 UTC，统一转本地，与会话标题一致）──────────
function pad2(n) { return String(n).padStart(2, "0"); }
function fmtDateTime(d) {
  return `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())} ${pad2(d.getHours())}:${pad2(d.getMinutes())}`;
}
function fmtTime(d) { return `${pad2(d.getHours())}:${pad2(d.getMinutes())}`; }

function listJsonl(dir) {
  let entries;
  try { entries = fs.readdirSync(dir); } catch { return []; }
  return entries
    .filter((f) => f.endsWith(".jsonl"))
    .map((f) => path.join(dir, f))
    .map((p) => ({ p, mtime: fs.statSync(p).mtimeMs }))
    .sort((a, b) => a.mtime - b.mtime); // 旧 → 新
}

// ── 项目发现 ─────────────────────────────────────────────────────────────────
function discoverProjects(projectsDir) {
  let dirs;
  try { dirs = fs.readdirSync(projectsDir, { withFileTypes: true }); } catch { return []; }

  const results = [];
  for (const ent of dirs) {
    if (!ent.isDirectory()) continue;
    const projDir = path.join(projectsDir, ent.name);
    const files = listJsonl(projDir);
    if (files.length === 0) continue;
    const latestMtime = Math.max(...files.map((f) => f.mtime));
    results.push({
      hash: ent.name,
      pathHint: guessProjectPath(files) || "(unknown)",
      latestMtime,
      sessionCount: files.length,
    });
  }
  results.sort((a, b) => b.latestMtime - a.latestMtime); // 最近活跃优先
  return results;
}

function guessProjectPath(files) {
  for (let i = files.length - 1; i >= 0; i--) { // 最新文件更可能有 cwd
    let lines;
    try { lines = fs.readFileSync(files[i].p, "utf-8").split("\n"); } catch { continue; }
    for (const line of lines) {
      const s = line.trim();
      if (!s) continue;
      let rec;
      try { rec = JSON.parse(s); } catch { continue; }
      const cwd = rec.cwd || rec.workingDirectory;
      if (cwd) return cwd;
    }
  }
  return null;
}

function printProjectList(projects) {
  console.log("\n📂 Claude Code 项目列表（按最近活跃排序）\n");
  console.log(`  ${"#".padEnd(4)} ${"最近活跃".padEnd(20)} ${"会话数".padEnd(6)} 项目路径`);
  console.log("  " + "─".repeat(75));
  projects.forEach((p, i) => {
    const dt = fmtDateTime(new Date(p.latestMtime));
    console.log(`  ${String(i + 1).padEnd(4)} ${dt.padEnd(20)} ${String(p.sessionCount).padEnd(6)} ${p.pathHint}`);
  });
  console.log("");
}

// ── 消息提取 ─────────────────────────────────────────────────────────────────
const MIN_TEXT_LENGTH = 20;
const NOISE_PREFIXES = [
  "Base directory for this skill",
  "<system-reminder",
  "<SUBAGENT-STOP",
  "<command-message",
  "<command-name",
  "<local-command",
  "<task-notification",
  "[Request interrupted",
  "Caveat:",
  "This session is being continued",
];
function isNoise(text) {
  const s = text.replace(/^\s+/, "");
  return NOISE_PREFIXES.some((p) => s.startsWith(p));
}

function extractTextContent(content) {
  if (typeof content === "string") {
    const s = content.trim();
    return s.length >= MIN_TEXT_LENGTH ? s : null;
  }
  if (Array.isArray(content)) {
    const parts = [];
    for (const block of content) {
      if (!block || typeof block !== "object") continue;
      const t = block.type || "";
      if (t === "tool_use" || t === "tool_result" || t === "thinking") continue;
      if (t === "text") {
        const txt = (block.text || "").trim();
        if (txt) parts.push(txt);
      }
    }
    const combined = parts.join("\n").trim();
    return combined.length >= MIN_TEXT_LENGTH ? combined : null;
  }
  return null;
}

function parseTimestamp(record) {
  for (const key of ["timestamp", "createdAt", "created_at", "ts"]) {
    const val = record[key];
    if (!val) continue;
    if (typeof val === "number") {
      return new Date(val > 1e10 ? val : val * 1000);
    }
    const d = new Date(String(val)); // ISO（含 Z）→ JS Date，本地格式化时即为本地时区
    if (!isNaN(d.getTime())) return d;
  }
  return null;
}

function extractFromJsonl(filepath) {
  const messages = [];
  const fileMtime = new Date(fs.statSync(filepath).mtimeMs);
  let lines;
  try { lines = fs.readFileSync(filepath, "utf-8").split("\n"); }
  catch (e) { process.stderr.write(`  ⚠️  读取 ${path.basename(filepath)} 时出错: ${e.message}\n`); return messages; }

  for (const line of lines) {
    const s = line.trim();
    if (!s) continue;
    let record;
    try { record = JSON.parse(s); } catch { continue; }

    const recordType = record.type;
    if (!["user", "assistant", "message"].includes(recordType)) continue;

    // 结构化过滤：子代理对话 / 系统元消息
    if (record.isSidechain || record.isMeta) continue;

    const msg = (record.message && typeof record.message === "object") ? record.message : record;
    const role = msg.role || record.role;
    if (!["user", "assistant", "human"].includes(role)) continue;

    const content = msg.content;
    if (content == null) continue;

    const text = extractTextContent(content);
    if (!text) continue;

    if ((role === "user" || role === "human") && isNoise(text)) continue;

    const ts = parseTimestamp(record) || fileMtime;
    messages.push({
      role: role === "user" || role === "human" ? "human" : "assistant",
      text,
      timestamp: ts,
    });
  }
  return messages;
}

function loadProjectSessions(projDir, since) {
  let files = listJsonl(projDir);
  if (since) files = files.filter((f) => f.mtime >= since.getTime());

  const all = [];
  for (const f of files) {
    const msgs = extractFromJsonl(f.p);
    if (msgs.length === 0) continue;
    const sessionId = path.basename(f.p, ".jsonl").slice(0, 8);
    // 会话起点用最早一条消息的真实时间戳，而非 mtime（resume 会让 mtime 失真）
    const sessionStart = new Date(Math.min(...msgs.map((m) => m.timestamp.getTime())));
    for (const m of msgs) { m.sessionId = sessionId; m.sessionStart = sessionStart; }
    all.push(...msgs);
  }
  all.sort((a, b) => a.timestamp - b.timestamp);
  return all;
}

// ── 骨架格式化 ───────────────────────────────────────────────────────────────
function formatSkeleton(messages, projectPath, maxChars = 0) {
  const lines = [
    `# 开发过程骨架 — ${projectPath}`,
    `_生成时间：${fmtDateTime(new Date())}_`,
    `_共 ${messages.length} 条对话消息_\n`,
  ];
  let current = null;
  for (const m of messages) {
    if (m.sessionId !== current) {
      current = m.sessionId;
      lines.push(`\n## 会话 ${m.sessionId}  ·  ${fmtDateTime(m.sessionStart)}\n`);
    }
    const roleLabel = m.role === "human" ? "👤 用户" : "🤖 Claude";
    let text = m.text;
    if (maxChars && text.length > maxChars) text = text.slice(0, maxChars) + "…（已截断）";
    lines.push(`**${roleLabel}** \`${fmtTime(m.timestamp)}\`\n${text}\n`);
  }
  return lines.join("\n");
}

// ── 参数解析 ─────────────────────────────────────────────────────────────────
function parseSince(str) {
  const m = /^(\d+)d$/.exec(str);
  if (m) return new Date(Date.now() - parseInt(m[1], 10) * 86400000);
  const d = new Date(str);
  return isNaN(d.getTime()) ? null : d;
}

function parseArgs(argv) {
  const args = { list: false, project: null, since: null, output: null, maxChars: 0 };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--list") args.list = true;
    else if (a === "--project") args.project = argv[++i];
    else if (a === "--since") args.since = argv[++i];
    else if (a === "--output") args.output = argv[++i];
    else if (a === "--max-chars") args.maxChars = parseInt(argv[++i], 10) || 0;
  }
  return args;
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const projectsDir = getClaudeProjectsDir();
  const projects = discoverProjects(projectsDir);

  if (projects.length === 0) {
    console.log(`❌ 未找到任何 Claude Code 项目，请确认 ${projectsDir} 存在`);
    process.exit(1);
  }

  if (args.list || !args.project) {
    printProjectList(projects);
    if (!args.project) console.log("👉 使用方式：node extract_sessions.js --project <序号或hash> [--since 7d]");
    return;
  }

  const target = String(args.project).trim();
  let projHash = null;
  if (/^\d+$/.test(target)) {
    const idx = parseInt(target, 10) - 1;
    if (idx >= 0 && idx < projects.length) projHash = projects[idx].hash;
    else { console.log(`❌ 序号 ${target} 超出范围（共 ${projects.length} 个项目）`); process.exit(1); }
  } else {
    const hit = projects.find((p) => p.hash.includes(target) || p.pathHint.includes(target));
    if (hit) projHash = hit.hash;
    else { console.log(`❌ 未找到匹配的项目：${target}`); process.exit(1); }
  }

  const projDir = path.join(projectsDir, projHash);
  const since = args.since ? parseSince(args.since) : null;
  console.log("\n⏳ 正在读取项目会话...");

  const messages = loadProjectSessions(projDir, since);
  if (messages.length === 0) {
    console.log("⚠️  未提取到任何有效消息（可能被 --since 过滤或文件格式不符）");
    process.exit(0);
  }

  const matched = projects.find((p) => p.hash === projHash) || {};
  const projectPath = matched.pathHint || projHash;
  const skeleton = formatSkeleton(messages, projectPath, args.maxChars);

  if (args.output) {
    fs.writeFileSync(args.output, skeleton, "utf-8");
    console.log(`✅ 骨架已写入：${args.output}`);
  } else {
    console.log(skeleton);
  }
}

main();
